packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/find.ts

```diff
@@ -266,13 +268,24 @@ export const performFind = async <T = unknown, A = unknown>(
   const migratedDocuments: Array<SavedObjectsFindResult<T>> = [];
   try {
     for (const savedObject of savedObjects) {
+      let migrated: SavedObjectUnsanitizedDoc | undefined;
       const { sort, score, ...rawObject } = savedObject;
-      const migrated = disableExtensions
-        ? migrationHelper.migrateStorageDocument(rawObject)
-        : await migrationHelper.migrateAndDecryptStorageDocument({
-            document: rawObject,
-            typeMap: redactTypeMap,
-          });
+
+      if (fields === undefined || fields.length === 0) {
+        // If this document only contains a subset of fields it's not possible to migrate it but we still try to decrypt it
+        migrated = await encryptionHelper.optionallyDecryptAndRedactSingleResult(
```

Comment:
@jeramysoucy this seems risky to me. Did refactoring allow decryption on partial fields?

Reply:
> we discussed the required AAD properties still being available in source
For partial updates, for example, that is true, but I think we overlooked this scenario of `find` with a partial field list.

I may be missing some details here of the specific use case/regression, but if the documents only contain a subset of fields, and is not guaranteed to contain the full list of AAD attributes, then we cannot be sure AAD can be constructed correctly for decryption.

Reply:
Before issuing the ES query, we could check if the type is encrypted, and augment the incoming field list with the AAD attributes. This would give us a much better chance of decrypting a partial object, only failing if the executing version does not know about some new future version AAD attribute...
However, it is best (correct?) practice to introduce attributes to AAD at least one version before populating them, so we could have high confidence in this approach if we need to implement it.

Reply:
yeah, but from what I understand `optionallyDecryptAndRedactSingleResult` will always resolve, if we couldn't decrypt then we'll strip the encrypted attributes but if we could then we'll return the unencrypted result

Reply:
You're right, it will allow a graceful error case, however, it will also still log the failure from the lower-level code: https://github.com/elastic/kibana/blob/ac04952af7446dbebf7f3ff7f4c11f74a5d0682b/x-pack/plugins/encrypted_saved_objects/server/crypto/encrypted_saved_objects_service.ts#L551

So it may be fine that we cannot decrypt, we just move on and expect the calling code to handle the non-decrypted case. But these will manifest as serverless alerts when this happens.

Reply:
> augment the incoming field list with the AAD attributes

I guess an even more optimal implementation would be to do this only if one of the requested fields are encrypted. Otherwise there's no point to even try to decrypt.

While this gives the safest behavior (we'll always be able to decrypt) it could potentially include a large amount of extra fields which the caller was unaware of. And here I think it's an accurate assumption that if a caller specifies a list of fields, they're concerned about the performance of retrieving all fields.

Sometimes this concern is purely a UI one, but in such a case they could decide to let the UI specify a fields list but not pass that list onto the saved objects client, but rather `_.pick` the list on the server-side.

I'm particularly thinking of Rules which have quite a long list of AAD attributes https://github.com/elastic/kibana/blob/dont-migrate-partial-docs/x-pack/plugins/alerting/server/saved_objects/index.ts#L50-L69

Reply:
> I guess an even more optimal implementation would be to do this only if one of the requested fields are encrypted. Otherwise there's no point to even try to decrypt.

++ to this

> While this gives the safest behavior (we'll always be able to decrypt) it could potentially include a large amount of extra fields which the caller was unaware of. And here I think it's an accurate assumption that if a caller specifies a list of fields, they're concerned about the performance of retrieving all fields.
> 
> Sometimes this concern is purely a UI one, but in such a case they could decide to let the UI specify a fields list but not pass that list onto the saved objects client, but rather _.pick the list on the server-side.

Could we push the decision up to the caller? Add an option to "allow decryption failure" or opposite, to "ensure decryption"? Default to the safe option, and allow callers to override with the less safe option to gain efficiency?

Reply:
I think it's worth considering a performance vs safe mode... We would still need to figure out a way not to log decryption failures in performance mode.

I can think of two alternatives:
1. Don't allow retrieving encrypted fields with the fields option at all. 
2. Don't allow retrieving encrypted fields with the fields option unless all the AAD fields are also included.

```
  const includesAttributesToDecrypt = fields.includes(type.attributesToEncrypt);
  const notAllAttributesIncluded = !attributesToIncludeInAAD.every((f) => fields.includes(f));

  if (includesAttributesToDecrypt && notAllAttributesIncluded) {
    throw new Error("You can't have your cake and eat it");
  }
```

That way encryption would never fail and we won't log our error/fire of alerts.

It feels like (2) is a bit brittle and if you add a new AAD field you could easily forget to update the calling site leading to unexpected runtime errors.

Reply:
> 1. Don't allow retrieving encrypted fields with the fields option at all.
> 2. Don't allow retrieving encrypted fields with the fields option unless all the AAD fields are also included.

I think option 1 could be viable if we don't have existing use cases where decrypted fields are needed.

> It feels like (2) is a bit brittle and if you add a new AAD field you could easily forget to update the calling site leading to unexpected runtime errors.

True, but it is not really more efficient than automatically including all the AAD fields, is it?. We could just add in any missing AAD fields and maybe log a warning?

> We would still need to figure out a way not to log decryption failures in performance mode.

We could pass a 'silent' or 'logging' argument down to the encryption service in the params object. We'd want to be careful that this would not get abused later though. Alternatively, we could still log, but change the log level or wording so that it doesn't get picked up by our alerting rule.

Reply:
@rudolf please update the Risk matrix with these considerations. It's important that they are made clear.

Reply:
From @jloleysens review:

> Would it be possible to view this PR as just restoring the previous behaviour of being able to .find over newer model versions.
> Addressing other shortcomings could be follow PRs since they seem to be increasing the complexity/risk somewhat of this change. WDYT @rudolf ? Is it better to resolve this in one PR?

I agree. This PR fixes the regression and restores the previous behavior. Where Kibana would currently completely fail to `find` with `fields` after a rollback it should now always be possible. We have identified a new risk which we previously overlooked and that is that after a rollback we could get a lot of encryption failure errors that could trigger an alert. But to address that there are difficult tradeoffs to make and it'd be better not to rush it. @jeramysoucy @azasypkin any objections?

Reply:
Actually this PR shouldn't change the decryption behavior at all since currently if we do a rollback we would perform a downgrade migration which would call decryption first before the migration function has a chance to fail https://github.com/elastic/kibana/blob/dont-migrate-partial-docs/packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/helpers/migration.ts#L89

So I don't think this PR should increase the chances that we get alerted on serverless or further increase any risks.

Reply:
@rudolf Not sure if the plan is to merge this PR as-is, but before we merge, should we handle, at a minimum, alternate/augmented logging text for when decryption fails in this case? Something simple, like `partialAttributes: boolean` in `DecryptParameters` would allow us to do this. If we surface this somehow in the logged message, we can then identify this case if/when it appears in serverless logs.

Reply:
Looks like our comments overlapped.

> Actually this PR shouldn't change the decryption behavior at all since currently if we do a rollback we would perform a downgrade migration which would call decryption first before the migration function

Just want to make sure I understand. In what event do we hit the decrypt on a partial ESO case? If there is still risk of hitting this case with an ESO that is missing an AAD attribute, we would want to be able to identify it.

Reply:
> So I don't think this PR should increase the chances that we get alerted on serverless or further increase any risks.

++ that's what I had in mind, all other considerations still valid

Reply:
This is based on my reading of the code and I haven't reproduced it in a test case.

Currently and if we merge this PR we could get decryption failure logs when:
 - SavedObjectsClient#find is called with `fields` where `fields` includes an encrypted attribute but not all AAD attributes. This might have been the case for a very long time and not zdt/serverless specific?
 - Since find usually strips the encrypted attributes, this would probably only apply to calls from the internal user

Unless the plugin expects to be able to read the encrypted field there should be no user impact. The only impact is noisy logging.

Reply:
Thanks! I'll open an issue for our team to investigate this/augment our logs.